### PR TITLE
schedule on

### PR DESCRIPTION
**Pull request by duke8253**

### CODE REVIEW

File: src/traffic_server/InkAPI.cc [hunk 1]
```diff
@@ -4489,6 +4489,56 @@ TSContScheduleEvery(TSCont contp, ink_hrtime every, TSThreadPool tp)
...
+TSAction
+TSContScheduleOn(TSCont contp, TSHRTime timeout, TSEventThread *event_thread)
+{
+  ink_release_assert(event_thread != nullptr);
+
+  sdk_assert(sdk_sanity_check_iocore_structure(contp) == TS_SUCCESS);
+
+  FORCE_PLUGIN_SCOPED_MUTEX(contp);
+
+  INKContInternal *i = (INKContInternal *)contp;
```
Comment (SolidWallOfCode) [inline src/traffic_server/InkAPI.cc]: No C-style casts!

File: src/traffic_server/InkAPI.cc [hunk 2]
```diff
@@ -4489,6 +4489,56 @@ TSContScheduleEvery(TSCont contp, ink_hrtime every, TSThreadPool tp)
...
+  TSAction action;
+
+  if (ink_atomic_increment((int *)&i->m_event_count, 1) < 0) {
+    ink_assert(!"not reached");
+  }
+
+  if (timeout == 0) {
+    action = reinterpret_cast<TSAction>(((EThread *)event_thread)->schedule_imm(i));
```
Comment (SolidWallOfCode) [inline src/traffic_server/InkAPI.cc]: No C-style casts! Also, don't double cast - create a local var instead.

File: iocore/eventsystem/UnixEThread.cc [hunk 3]
```diff
@@ -208,8 +208,10 @@ EThread::execute_regular()
   static EventMetrics METRIC_INIT;
 
   // give priority to immediate events
+  has_event_loop = true;
   for (;;) {
     if (unlikely(shutdown_event_system == true)) {
+      has_event_loop = false;
```
Comment (SolidWallOfCode) [inline iocore/eventsystem/UnixEThread.cc]: I'm dubious this is a good idea. I need to think about that.

File: src/traffic_server/InkAPI.cc (hunk 1 above)
Comment (duke8253) [reply src/traffic_server/InkAPI.cc]: This is copied from `TSContSchedule`, should I change those as well?

**Comment by SolidWallOfCode** [comment]
"Schedule on my wayward thread.
There'll be peace when you dead.
Lay your weary code to rest.
Don't you run no more."

**Comment by SolidWallOfCode** [comment]
I think a bit of clean up and you can propose this to the dev list.

File: doc/developer-guide/api/functions/TSContSchedule.en.rst [hunk 4]
```diff
@@ -26,21 +26,20 @@ Synopsis
...
 
 Description
 ===========
 
-Schedules :arg:`contp` to run :arg:`delay` milliseconds in the future. This is approximate. The delay
-will be at least :arg:`delay` but possibly more. Resultions finer than roughly 5 milliseconds will
-not be effective. :arg:`contp` is required to have a mutex, which is provided to
-:func:`TSContCreate`.
+Schedules :arg:`contp` to run :arg:`timeout` milliseconds in the future. This is approximate. The delay
+will be at least :arg:`timeout` but possibly more. Resultions finer than roughly 5 milliseconds will
```
[the remark this answers is not in the document]
Comment (duke8253) [reply doc/developer-guide/api/functions/TSContSchedule.en.rst]: Hmm, I think it should be "resolutions" here.

File: doc/developer-guide/api/functions/TSContSchedule.en.rst [hunk 5]
```diff
@@ -26,21 +26,20 @@ Synopsis
 
 `#include <ts/ts.h>`
 
-.. function:: TSAction TSContSchedule(TSCont contp, ink_hrtime delay, TSThreadPool tp)
+.. function:: TSAction TSContSchedule(TSCont contp, TSHRTime timeout, TSThreadPool tp)
```
[the remark this answers is not in the document]
Comment (duke8253) [reply doc/developer-guide/api/functions/TSContSchedule.en.rst]: `ink_hrtime` is `int64`, I can replace it with that.

**Comment by oschaaf** [comment]
I think this looks nice! I don't want to bloat this, but I did wonder about the following earlier:
Would it make sense to be able to align continuations to other continuations? 

So a plugin author can specify he wants to run code associated to *this* continuation on whatever thread code associated to *that* continuation is running?

**Comment by duke8253** [comment]
@oschaaf I think it's possible, but seems like it should be a seperate PR.

**Comment by SolidWallOfCode** [comment]
I don't see why you can't do what @oschaaf asked for with this PR. You can grab the affine thread from Continuation A and then make Continuation B affine to the same thread. Is there something more than that desired?

**Comment by oschaaf** [comment]
@SolidWallOfCode 
The way I was thinking about it, when specifying code execution alignment based at the continuation level, the alignment would persist. So when Continuation A later on switches thread alignment for some reason, Continuation B would automatically follow.

I was hoping that in practice a plugin would be (easily) able to align its own state machine with another one, usually one controlled by ATS (e.g. the http state machine). I think that could be useful, but this is probably easier said then done... 

I thought it would be a good to share the idea because in the event we do want to have something like that in the future we may want to keep that in mind when mulling over the API's we are about to introduce here.

**Comment by SolidWallOfCode** [comment]
The complication is plugins don't have state machines, continuations do. That is, for a given plugin Alpha, there can be an arbitrarily large number of state machines for Alpha in different continuations. If there is another plugin Bravo that wants its continuations to align with those of Alpha, it's an Bravo -> Alpha mapping that must be maintained, but N mappings for each pair (?) of continuations (Continuation-Bravo-K -> Continuation-Alpha-K) for K in 0..N-1. That would be challenging. I think it would be possible with this API, however, if each Bravo Continuation cooperated with its corresponding Alpha continuation. Then the Alpha continuation could send a message to the corresponding Bravo continuation if it engages in thread switching. This means Alpha and Bravo continuations have to cooperate at least to the extent of tracking each other's existence. I can't think of an actual use case, so it seems a bit obscure to support explicitly in the core.

**Comment by oschaaf** [comment]
Well, yeah , I thought that this might escalate quickly :) hence it’s probably best considered out of scope for this PR, which is I feel is really good to have as-is. For the aligning at the continuation level I think you are right in that inter-plugin alignment is probably rarely needed. But how about aligning a plug-in to ATS internals?  (Ps - maybe we should take this elsewhere? I don’t mean to introduce a lot of noise to handling this PR).

**Pull request closed without merging**